node/pkg/watchers/evm/blocks_by_timestamp.go

```diff
@@ -0,0 +1,157 @@
+package evm
+
+import (
+	"sort"
+	"sync"
+
+	"go.uber.org/zap"
+)
+
+const (
+	BTS_MAX_BLOCKS = 10000
+)
+
+type (
+	BlocksByTimestamp struct {
+		// cache is ordered by timestamp, blockNum. There may be multiple entries for the same timestamp, but not the same block.
+		cache Blocks
+
+		// maxCacheSize is used to trim the cache.
+		maxCacheSize int
+
+		// mutex is used to protect the cache.
+		mutex sync.Mutex
+	}
+
+	Blocks []Block
+
+	Block struct {
+		Timestamp uint64
+		BlockNum  uint64
+	}
+)
+
+// NewBlocksByTimestamp creates an empty cache of blocks by timestamp.
+func NewBlocksByTimestamp(maxCacheSize int) *BlocksByTimestamp {
+	return &BlocksByTimestamp{
+		cache:        Blocks{},
+		maxCacheSize: maxCacheSize,
+	}
+}
+
+// AddLatest adds a block to the end of the cache. This is meant to be used in the normal scenario when a new latest block is received. If the specified
+// timestamp or block number is less than the latest in the cache (most likely a rollback), the cache will be truncated and the new value inserted.
+func (bts *BlocksByTimestamp) AddLatest(logger *zap.Logger, timestamp uint64, blockNum uint64) {
+	bts.mutex.Lock()
+	defer bts.mutex.Unlock()
+	l := len(bts.cache)
+	if l > 0 && (blockNum <= bts.cache[l-1].BlockNum || timestamp < bts.cache[l-1].Timestamp) {
+		for idx := len(bts.cache) - 1; idx >= 0; idx-- {
+			if bts.cache[idx].BlockNum < blockNum && bts.cache[idx].Timestamp <= timestamp {
+				logger.Warn("rollback detected in timestamp cache",
+					zap.Uint64("oldLatestBlockNum", bts.cache[len(bts.cache)-1].BlockNum),
+					zap.Uint64("oldLatestTimestamp", bts.cache[len(bts.cache)-1].Timestamp),
+					zap.Uint64("newLatestBlockNum", blockNum),
+					zap.Uint64("newLatestTimestamp", timestamp),
+				)
+				bts.cache = bts.cache[:idx+1]
```

Comment:
I think in the case where the cache is fresh and a rollback is greater than `l-1`, the cache won't get pruned and will no longer be sorted appropriately. :cry:

Reply:
Fixed and wrote a test to confirm it.